### ISSUE TITLE
False positive E0308 with associated types, `Fn` and tuples.

### ISSUE THREAD

**Issue by Ddystopia** [rust-lang/rust#135727]
I tried this code:

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=trait+Trait%3A+Sized+%7B%0A++++type+Assoc%3B%0A++++fn+foo%28self%2C+f%3A+impl+FnOnce%28Self%3A%3AAssoc%29+-%3E+Self%3A%3AAssoc%29%3B%0A%7D%0A%0Aimpl%3C%27a%2C+T1%3E+Trait+for+%28T1%2C%29+%7B%0A++++type+Assoc+%3D+%28T1%2C%29%3B%0A++++fn+foo%28self%2C+_%3A+impl+FnOnce%28Self%3A%3AAssoc%29+-%3E+Self%3A%3AAssoc%29+%7B%0A++++++++unimplemented%21%28%29%0A++++%7D%0A%7D%0A%0Aimpl%3C%27a%2C+T1%2C+T2%3E+Trait+for+%28T1%2C+T2%29+%7B%0A++++type+Assoc+%3D+%28T1%2C+T2%29%3B%0A++++fn+foo%28self%2C+f%3A+impl+FnOnce%28Self%3A%3AAssoc%29+-%3E+Self%3A%3AAssoc%29+%7B%0A++++++++let+%28x%2C+y%29+%3D+self%3B%0A++++++++%28y%2C%29.foo%28%7C%28y%2C%29%7C+f%28%28x%2C+y%29%29%29%0A++++%7D%0A%7D%0A)

```rust
trait Trait: Sized {
    type Assoc;
    fn foo(self, f: impl FnOnce(Self::Assoc) -> Self::Assoc);
}

impl<'a, T1> Trait for (T1,) {
    type Assoc = (T1,);
    fn foo(self, _: impl FnOnce(Self::Assoc) -> Self::Assoc) {
        unimplemented!()
    }
}

impl<'a, T1, T2> Trait for (T1, T2) {
    type Assoc = (T1, T2);
    fn foo(self, f: impl FnOnce(Self::Assoc) -> Self::Assoc) {
        let (x, y) = self;
        (y,).foo(|(y,)| f((x, y)))
    }
}
```

I expected to see this happen: code should compile
Instead, this happened: code did not compile
```rust
error[E0308]: mismatched types
  --> src/main.rs:17:25
   |
17 |         (y,).foo(|(y,)| f((x, y)))
   |                         ^^^^^^^^^ expected a tuple with 1 element, found one with 2 elements
   |
   = note: expected tuple `(T2,)`
              found tuple `(T1, T2)`
```

`f` has type `FnOnce(Self::Assoc) -> Self::Assoc`, which is `FnOnce((T1, T2)) -> (T1, T2)`. But compiler wants `(T2,)` as an argument. This is not a soundness issue, as if I give it `(y,)` instead of `(x, y)` it will give correct error, that it expects `(T1, T2)`.

### Meta

`rustc --version --verbose`:
```
rustc 1.85.0-nightly (8742e0556 2024-12-28)
binary: rustc
commit-hash: 8742e0556dee3c64f7144de2fb2e88936418865a
commit-date: 2024-12-28
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.6
```

Here too:

`rustc --version --verbose`:
```
rustc 1.86.0-nightly (049355708 2025-01-18)
binary: rustc
commit-hash: 049355708383ab1b9a1046559b9d4230bdb3a5bc
commit-date: 2025-01-18
host: x86_64-unknown-linux-gnu
release: 1.86.0-nightly
LLVM version: 19.1.7
```

(It reproduces with `cargo +stable check` too)

**Comment by Ddystopia** [comment]
Whoops, I misunderstood the error, it is actually correct
